### PR TITLE
File generating function

### LINKED ISSUE #2: Create a file generating function
We want to use a pseudo-random byte generator using math/rand in go to generate file contents

### PR DESCRIPTION
**Pull request by alicjakwie**
closes #2

### CODE REVIEW

File: alicja/sharedgo/filegen.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
+package main
```
Comment (cianbgoog) [inline alicja/sharedgo/filegen.go]: You don't want to be in package main. Again, I'd probably use a package utility.

File: alicja/sharedgo/filegen.go [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+import (
+	"fmt"
+	// "math/rand"
+	"strconv"
+	"crypto/sha1"
+)
+type HermesFile struct {
+	name string
+	contents string
```
Comment (cianbgoog) [inline alicja/sharedgo/filegen.go]: This would normally be an `io.Reader` in go

File: alicja/sharedgo/filegen.go [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+}
+func generate_file_name (file_ID int, file_checksum string) string{
```
Comment (cianbgoog) [inline alicja/sharedgo/filegen.go]: I think this should probably be a method of `HermesFile`

File: alicja/sharedgo/filegen.go [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+	var id_string string=""
+	if file_ID < 10 {
+		id_string = "0" + strconv.Itoa(file_ID)
+	} else {
+		id_string = strconv.Itoa(file_ID)
+	}
+	var file_name string = "Hermes_"+id_string+file_checksum
+	return file_name 
+}
+func generate_file_contents (file_ID int) string{
```
Comment (cianbgoog) [inline alicja/sharedgo/filegen.go]: Make this a method of `HermesFile`, and ideally have it update the `HermesFile` you have, to create an io.Reader, than then a `string` (ideally, you probably want `[]Byte` rather `string` in any case - this data isn't really semantically meaningful, so you probably don't want a `string` type).

File: alicja/sharedgo/filegen.go [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+	contents string
+}
+func generate_file_name (file_ID int, file_checksum string) string{
+	var id_string string=""
+	if file_ID < 10 {
+		id_string = "0" + strconv.Itoa(file_ID)
+	} else {
+		id_string = strconv.Itoa(file_ID)
+	}
+	var file_name string = "Hermes_"+id_string+file_checksum
```
Comment (cianbgoog) [inline alicja/sharedgo/filegen.go]: You could I think avoid the dance above by going `fmt.Sprintf("Hermes_%2d_%v", file_ID, file_checksum)`

File: alicja/sharedgo/filegen.go [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+	return file_name 
+}
+func generate_file_contents (file_ID int) string{
+	return "jhfvjhdfjhfjjhjhdfvjvcvfjh";
+}
+func generate_file_checksum (file HermesFile) string{
```
Comment (cianbgoog) [inline alicja/sharedgo/filegen.go]: I suspect you should use `https://golang.org/pkg/crypto/sha1/` so that you can checksum the file bit by bit, rather than all in one go. You'll at least want to do this when you're reading files to verify them.

Also, if you make this a method on HermesFile, then you don't need to pass in any args to this.

File: alicja/sharedgo/filegen.go [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+	return checksum;
+}
+func generate_file (id int) HermesFile {
+	file := HermesFile{}
+	file.contents = generate_file_contents(id)
+	var checksum string = generate_file_checksum(file)
+	file.name = generate_file_name(id, checksum)
+	return file
+}
+func main(){
```
Comment (cianbgoog) [inline alicja/sharedgo/filegen.go]: No need for main - if you need to make sure the contents work, a unit test might be better?

**Review by cianbgoog — CHANGES_REQUESTED** [review]
Left some comments - in general I think these should probably all be methods on `HermesFile`, and the golang style is normally to use CameCase rather than snake_case. You also probably don't want any strings here, since you're only reading, writing and diffing this data - so you can't care about e.g. encoding, and in general for this kind of data, golang would generally prefer `io.Reader`s and `io.Writer`s rather than raw strings.

File: hermes/file_gen.go [hunk 8]
```diff
@@ -0,0 +1,42 @@
+package main
+import (
```
Comment (evanSpendlove) [inline hermes/file_gen.go:16]: Add whitespace.
```suggestion

import (
```

File: hermes/file_gen.go [hunk 9]
```diff
@@ -0,0 +1,42 @@
+package main
```
Comment (evanSpendlove) [inline hermes/file_gen.go]: Change the package to something other than main, ideally `common` or `shared` as this code will be common to/shared between all file systems.

Also, add a description of the package as a comment.

Comment (evanSpendlove) [reply hermes/file_gen.go]: Also, add the license header, the author line and a description of the file as comments.

File: hermes/file_gen.go [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+import (
+	"fmt"
+	// "math/rand"
+	"strconv"
+	"crypto/sha1"
+)
+type HermesFile struct {
```
Comment (evanSpendlove) [inline hermes/file_gen.go:21]: Describe the HermesFile struct in a comment here.

File: hermes/file_gen.go [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+	name string
+	contents string
+}
+func generate_file_name (file_ID int, file_checksum string) string{
```
Comment (evanSpendlove) [inline hermes/file_gen.go]: Add a new line to separate from the HermesFile struct.

Add a comment explaining what this function does and any nuances about its use/arguments.

File: hermes/file_gen.go [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+	var id_string string=""
```
Comment (evanSpendlove) [inline hermes/file_gen.go]:
```suggestion
	id_string := ""
```

File: hermes/file_gen.go [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+	if file_ID < 10 {
```
Comment (evanSpendlove) [inline hermes/file_gen.go]: Add a comment explaining what IDs < 10 need the extra 0 (explain the fixed filename length consistency)

File: hermes/file_gen.go [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+		id_string = "0" + strconv.Itoa(file_ID)
+	} else {
+		id_string = strconv.Itoa(file_ID)
+	}
+	var file_name string = "Hermes_"+id_string+file_checksum
```
Comment (evanSpendlove) [inline hermes/file_gen.go]: Using := instead of var = is more consise.

```suggestion
	file_name string := "Hermes_" + id_string + file_checksum
```

File: hermes/file_gen.go [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+	return file_name 
+}
+func generate_file_contents (file_ID int) string{
```
Comment (evanSpendlove) [inline hermes/file_gen.go]: Add a comment explaining what this function does and any nuances about its use/arguments.

File: hermes/file_gen.go [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+	return "jhfvjhdfjhfjjhjhdfvjvcvfjh";
+}
+func generate_file_checksum (file HermesFile) string{
```
Comment (evanSpendlove) [inline hermes/file_gen.go]: Add a comment explaining what this function does and any nuances about its use/arguments.

File: hermes/file_gen.go [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+	
+	file_contents := []byte(file.contents)
+	hash := sha1.Sum(file_contents)
+	checksum := fmt.Sprintf("%s%x","_", hash)
+	return checksum;
+}
+func generate_file (id int) HermesFile {
```
Comment (evanSpendlove) [inline hermes/file_gen.go]: Add a comment explaining what this function does and any nuances about its use/arguments.

File: hermes/file_gen.go [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+	file := HermesFile{}
+	file.contents = generate_file_contents(id)
+	var checksum string = generate_file_checksum(file)
+	file.name = generate_file_name(id, checksum)
+	return file
+}
+func main(){
```
Comment (evanSpendlove) [inline hermes/file_gen.go]: Remove the main function.

File: alicja/sharedgo/filegen.go [hunk 19]
```diff
@@ -10,7 +10,12 @@ type HermesFile struct {
 	contents string
 }
 func generate_file_name (file_ID int, file_checksum string) string{
-	var id_string string = strconv.Itoa(file_ID)
+	var id_string string=""
```
Comment (deniadelcid) [inline alicja/sharedgo/filegen.go]:
```suggestion
	id_string := ""
```

File: hermes/file_gen.go (hunk 17 above)
Comment (evanSpendlove) [reply hermes/file_gen.go]: None of these functions are exported from this file (i.e. none are public), these are all private functions.

At least one should be made public/exported by capitalising the function name.

Also, I think we are using camelcase for function names as per convention.

File: alicja/sharedgo/filegen.go (hunk 5 above)
Comment (alicjakwie) [reply alicja/sharedgo/filegen.go]: Thank you for the suggestion! Tried this out but it leaves a space in the name so for example for file_ID = 3 we get  Hermes_ 3_checksum and am a bit worried about spaces in the name

Comment (cianbgoog) [reply alicja/sharedgo/filegen.go]: Try fmt.Sprintf("Hermes_%02d_%v", file_ID, file_checksum")

Comment (cianbgoog) [reply alicja/sharedgo/filegen.go]: https://play.golang.org/p/-cvaVQM78nq

File: hermes/file_gen.go [hunk 20]
```diff
@@ -0,0 +1,38 @@
...
+}
+func (file *HermesFile) GenerateFileName(file_ID int, file_checksum string) {
+	id_string :=""
+	if file_ID < 10 {
+		id_string = "0" + strconv.Itoa(file_ID)
+	} else {
+		id_string = strconv.Itoa(file_ID)
+	}
+	file_name := "Hermes_"+id_string+file_checksum
+	file.name = file_name
```
Comment (evanSpendlove) [inline hermes/file_gen.go]: Can simplify this.
```suggestion
	file.name := "Hermes_"+id_string+file_checksum
```

File: hermes/file_gen.go [hunk 21]
```diff
@@ -0,0 +1,38 @@
...
+}
+func (file *HermesFile) GenerateFileContents(file_ID int) {
+	file.contents = "jhfvjhdfjhfjjhjhdfvjvcvfjh";
+}
+func (file HermesFile) GenerateFileChecksum() string {
+
```
Comment (evanSpendlove) [inline hermes/file_gen.go]: Remove whitespace.
```suggestion
```

File: hermes/file_gen.go [hunk 22]
```diff
@@ -0,0 +1,38 @@
+package hermes
```
Comment (evanSpendlove) [inline hermes/file_gen.go:14]: Add comment with license header, author, and short description of the purpose of this file.

**Review by evanSpendlove — CHANGES_REQUESTED** [review]
- [x] Add more comments detailing what each function does
- [x] Export one or more of the methods by capitalising it.
- [x] Convert function names to CamelCase over snake_case. 
- [x] Other minor code changes

-- Edit --
I think some of these may have been mentioned by Cian, so apologies for duplication.

File: hermes/file_gen.go [hunk 23]
```diff
@@ -1,28 +1,28 @@
-package main
+package hermes
+
 import (
 	"fmt"
-	// "math/rand"
 	"strconv"
 	"crypto/sha1"
 )
 type HermesFile struct {
```
Comment (deniadelcid) [inline hermes/file_gen.go:21]: Just add some whitespace here.
```suggestion

type HermesFile struct {
```

File: hermes/file_gen.go [hunk 24]
```diff
@@ -1,28 +1,28 @@
...
 	name string
 	contents string
 }
-func generate_file_name (file_ID int, file_checksum string) string{
-	var id_string string=""
+func (file *HermesFile) GenerateFileName(file_ID int, file_checksum string) {
```
Comment (deniadelcid) [inline hermes/file_gen.go]: - I don't think that the file name generating function should be made public.

- nit: if you use snake_case, all your alphabetic characters should be in lower case. Ditto for all other similar situations in the rest of the file.

```suggestion

func (file *HermesFile) generateFileName(file_id int, file_checksum string) {
```

File: hermes/file_gen.go [hunk 25]
```diff
@@ -1,28 +1,28 @@
...
+	id_string :=""
```
Comment (deniadelcid) [inline hermes/file_gen.go]: nit leave one blank space on each side of your operators.
```suggestion
	id_string := ""
```

File: hermes/file_gen.go [hunk 26]
```diff
@@ -1,28 +1,28 @@
...
 	if file_ID < 10 {
 		id_string = "0" + strconv.Itoa(file_ID)
 	} else {
 		id_string = strconv.Itoa(file_ID)
 	}
-	var file_name string = "Hermes_"+id_string+file_checksum
-	return file_name 
+	file_name := "Hermes_"+id_string+file_checksum
```
Comment (deniadelcid) [inline hermes/file_gen.go]: - nit: leave a space around your operators
- use fmt.Sprintf() to concatenate strings instead of the + operator.

File: hermes/file_gen.go [hunk 27]
```diff
@@ -1,28 +1,28 @@
...
+	file.name = file_name
 }
-func generate_file_contents (file_ID int) string{
-	return "jhfvjhdfjhfjjhjhdfvjvcvfjh";
+func (file *HermesFile) GenerateFileContents(file_ID int) {
```
Comment (deniadelcid) [inline hermes/file_gen.go]: - nit: leave a blank line between functions.
- Generating file contents also shouldn't be public.
```suggestion

func (file *HermesFile) generateFileContents(file_ID int) {
```

File: hermes/file_gen.go [hunk 28]
```diff
@@ -1,28 +1,28 @@
...
+	file.contents = "jhfvjhdfjhfjjhjhdfvjvcvfjh";
 }
-func generate_file_checksum (file HermesFile) string{
+func (file HermesFile) GenerateFileChecksum() string {
```
Comment (deniadelcid) [inline hermes/file_gen.go]:
```suggestion

func (file HermesFile) GenerateFileChecksum() string {
```

File: hermes/file_gen.go [hunk 29]
```diff
@@ -1,28 +1,28 @@
...
 	
 	file_contents := []byte(file.contents)
 	hash := sha1.Sum(file_contents)
```
Comment (deniadelcid) [inline hermes/file_gen.go]:
```suggestion
	hash := sha1.Sum([]byte(file.contents))
```

File: hermes/file_gen.go [hunk 30]
```diff
@@ -0,0 +1,38 @@
...
+	file.name = file_name
+}
+func (file *HermesFile) GenerateFileContents(file_ID int) {
+	file.contents = "jhfvjhdfjhfjjhjhdfvjvcvfjh";
+}
+func (file HermesFile) GenerateFileChecksum() string {
+	
+	file_contents := []byte(file.contents)
+	hash := sha1.Sum(file_contents)
+	checksum := fmt.Sprintf("%s%x","_", hash)
```
Comment (deniadelcid) [inline hermes/file_gen.go]:
```suggestion
	checksum := fmt.Sprintf("_%x", hash)
```

File: hermes/file_gen.go [hunk 31]
```diff
@@ -0,0 +1,38 @@
...
+	return checksum;
+}
+func generate_file (id int) HermesFile {
```
Comment (deniadelcid) [inline hermes/file_gen.go]: This one should be exported (a.k.a made public).
```suggestion
func GenerateHermesFile (id int) HermesFile {
```

**Review by deniadelcid — COMMENTED** [review]
Left a few nits, some/most should get fixed by running gofmt.

File: hermes/file_gen.go [hunk 32]
```diff
@@ -1,40 +1,33 @@
...
 }
-func (file *HermesFile) GenerateFileName(file_ID int, file_checksum string) {
-	id_string := ""
-	if file_ID < 10 {
-		id_string = "0" + strconv.Itoa(file_ID)
-	} else {
-		id_string = strconv.Itoa(file_ID)
-	}
-	file.name := "Hermes_"+id_string+file_checksum
+func (file *HermesFile) generateFileName(file_ID int, file_checksum string) {
```
Comment (deniadelcid) [inline hermes/file_gen.go]: Use snake_case for variable names.
```suggestion
func (file *HermesFile) generateFileName(file_id int, file_checksum string) {
```

File: hermes/file_gen.go [hunk 33]
```diff
@@ -1,40 +1,33 @@
...
+	file_name :=  fmt.Sprintf("Hermes_%02d%v", file_ID, file_checksum)
+	file.name = file_name
```
Comment (deniadelcid) [inline hermes/file_gen.go]: - [ ] There is no need to declare intermediate variables whose sole role is to be assigned to another struct field. Just assign to the struct field directly.

- [ ] Just for a bit of parallelism with the file_id, lets prepend the `_` for the checksum here.

```suggestion
	file.name :=  fmt.Sprintf("Hermes_%02d_%v", file_id, file_checksum)
```

File: hermes/file_gen.go [hunk 34]
```diff
@@ -1,40 +1,33 @@
...
 }
-
 func (file *HermesFile) generateFileContents(file_ID int) {
```
Comment (deniadelcid) [inline hermes/file_gen.go]: - [x] Ditto. (Use snake_case for your variables)

- [x] Leave a blank line in between functions.

```suggestion

func (file *HermesFile) generateFileContents(file_ID int) {
```

File: hermes/file_gen.go (hunk 30 above)
Comment (deniadelcid) [reply hermes/file_gen.go]: If you are concatenating a constant string into another string, it just makes sense to make it part of the main string, it's less confusing.

File: hermes/file_gen.go (hunk 32 above)
Comment (deniadelcid) [inline hermes/file_gen.go]: Leave a blank line in between functions.

```suggestion

func (file *HermesFile) generateFileName(file_ID int, file_checksum string) {
```

File: hermes/file_gen.go [hunk 35]
```diff
@@ -1,40 +1,33 @@
...
 }
-
-func (file HermesFile) GenerateFileChecksum() string {
+func (file HermesFile) generateFileChecksum() string {
+	
 	file_contents := []byte(file.contents)
 	hash := sha1.Sum(file_contents)
-	checksum := fmt.Sprintf("_%x", hash)
+	checksum := fmt.Sprintf("%s%x","_", hash)
 	return checksum;
```
Comment (deniadelcid) [inline hermes/file_gen.go]: How about we return just the checksum and prepend the `_` inside generateFileName()?
```suggestion
        // return checksum in hex notation
	return fmt.Sprintf("%x", hash)
```

File: hermes/file_gen.go (hunk 30 above)
Comment (deniadelcid) [reply hermes/file_gen.go]: actually ignore this comment and lets proceed with only returning the checksum (without prepending anything here - see next comment).

File: hermes/file_gen.go [hunk 36]
```diff
@@ -1,40 +1,33 @@
...
 }
-func GenerateHermesFile (id int) HermesFile {
+func GenerateFile (id int) HermesFile {
```
Comment (deniadelcid) [inline hermes/file_gen.go]: I liked GenerateHermesFile better, why the change?

**Review by deniadelcid — COMMENTED** [review]
- [x] Please apply the general style comments from the suggestions in the `file_gen_test.go` file.

We're getting close to submitting, left a few minor comments.

File: hermes/file_gen.go [hunk 37]
```diff
@@ -0,0 +1,33 @@
+package hermes
+
+import (
+	"fmt"
+	"crypto/sha1"
+)
+
+type HermesFile struct {
+	name string
```
Comment (cianbgoog) [inline hermes/file_gen.go]: Should `Name` and `Contents` be exported?

File: hermes/file_gen.go [hunk 38]
```diff
@@ -10,21 +23,26 @@ type HermesFile struct {
...
 }
+
+// method of HermesFile generates the checksum of the file contents
 func (file HermesFile) generateFileChecksum() string {
-	
 	file_contents := []byte(file.contents)
 	hash := sha1.Sum(file_contents)
-	checksum := fmt.Sprintf("%s%x","_", hash)
-	return checksum;
+	// return checksum in hex notation
```
Comment (cianbgoog) [inline hermes/file_gen.go:40]: We don't need this comment  - in general you should use comments to explain why you do something, now what you're doing.

File: hermes/file_gen.go (hunk 37 above)
Comment (alicjakwie) [reply hermes/file_gen.go]:
```suggestion
	Name string
```

File: hermes/file_gen.go (hunk 38 above)
Comment (alicjakwie) [reply hermes/file_gen.go:40]: @deniadelcid suggested this comment

File: hermes/file_gen.go [hunk 39]
```diff
@@ -19,7 +19,7 @@ import (
 )
 
 type HermesFile struct {
-	name string
+	Name string
 	contents string
```
Comment (alicjakwie) [inline hermes/file_gen.go:23]:
```suggestion
	Contents string
```

File: hermes/file_gen.go [hunk 40]
```diff
@@ -0,0 +1,51 @@
...
+// method of HermesFile generates HermesFile.contents now a string without any significance in the future a pseudo random byte generator will be used
+func (file *HermesFile) generateFileContents(file_id int) {
+	file.contents = "jhfvjhdfjhfjjhjhdfvjvcvfjh";
+}
+
+// method of HermesFile generates the checksum of the file contents
+func (file HermesFile) generateFileChecksum() string {
+	file_contents := []byte(file.contents)
+	hash := sha1.Sum(file_contents)
+	// return checksum in hex notation
```
Comment (alicjakwie) [inline hermes/file_gen.go:40]:
```suggestion
```

File: hermes/file_gen.go [hunk 41]
```diff
@@ -0,0 +1,50 @@
...
+)
+
+type HermesFile struct {
+	Name string
+	Contents string
+}
+
+// method of HermesFile generates HermesFile.name of the form Hermes_id_checksum where id is an integer & id <= 50
+func (file *HermesFile) generateFileName(file_id int, file_checksum string) {
+	file.name = file.name :=  fmt.Sprintf("Hermes_%02d_%v", file_id, file_checksum)
```
Comment (cianbgoog) [inline hermes/file_gen.go]: I suspect you only want one file.name here :-P

Comment (alicjakwie) [reply hermes/file_gen.go]:
```suggestion
	file.name :=  fmt.Sprintf("Hermes_%02d_%v", file_id, file_checksum)
```

File: hermes/file_gen.go [hunk 42]
```diff
@@ -0,0 +1,50 @@
...
+
+import (
+	"fmt"
+	"crypto/sha1"
+)
+
+type HermesFile struct {
+	Name string
+	Contents string
+}
```
Comment (deniadelcid) [inline hermes/file_gen.go:24]: run gofmt on this file so that lines 22 and 23 are properly formatted/spaced.

**Review by deniadelcid — APPROVED** [review]
LGTM! Just run gofmt on the file before merging.

**Review by cianbgoog — APPROVED** [review]
